File: server/routes/tasks.py
from flask import request, jsonify
from flask_jwt_extended import jwt_required, get_jwt_identity
from models import Task, db
from models.task_utils import (
    add_task, get_root_tasks, get_task_with_subtasks,
    delete_task, move_subtask, toggle_task_completion,
    get_tasks_stats_by_date_range, get_tasks_with_filters
)
from models.task_dependency import (
    add_dependency, remove_dependency, get_user_dependencies, get_task_dependencies
)
from . import tasks_bp
from datetime import datetime
from socket_events import emit_task_created, emit_task_updated, emit_task_deleted, emit_task_completed

@tasks_bp.route('/', methods=['GET'])
@jwt_required()
def get_tasks():
    from sqlalchemy import or_, and_
    
    user_id = get_jwt_identity()
    date_str = request.args.get('date')
    
    # Use client's today if provided, otherwise fall back to server's today
    client_today_str = request.args.get('client_today')
    if client_today_str:
        try:
            today = datetime.strptime(client_today_str, '%Y-%m-%d').date()
        except ValueError:
            today = datetime.now().date()
    else:
        today = datetime.now().date()
    
    # Get tasks filtered by date if provided
    query = Task.query.filter_by(user_id=user_id)
    
    if date_str:
        try:
            # Parse date without timezone information - treats date as local timezone
            filter_date = datetime.strptime(date_str, '%Y-%m-%d').date()
            
            # A task is active on a date based on these rules:
            # - Tasks WITH deadline: Active from creation_date to deadline
            # - Tasks WITHOUT deadline (not completed): Only active on TODAY (rolls forward daily)
            # - Tasks WITHOUT deadline (completed): No longer active
            query = query.filter(
                db.func.date(Task.creation_date) <= filter_date,
                or_(
                    # Tasks with deadline - show in their date range
                    and_(
                        Task.deadline != None,
                        db.func.date(Task.deadline) >= filter_date
                    ),
                    # Tasks without deadline and not completed - only show on today
                    and_(
                        Task.deadline == None,
                        Task.completed == False,
                        filter_date == today
                    )
                )
            )
        except ValueError:
            return jsonify({'error': 'Invalid date format. Use YYYY-MM-DD without timezone'}), 400
    
    # Get all tasks for the date and build hierarchy
    tasks = query.all()
    root_tasks = [t for t in tasks if t.parent_id is None]
    return jsonify([get_task_with_subtasks(db.session, task.id, user_id) for task in root_tasks])


@tasks_bp.route('/', methods=['POST'])
@jwt_required()
def create_task():
    user_id = get_jwt_identity()
    data = request.get_json()

    if not data.get('name'):
        return jsonify({'success': False, 'message': 'Task name is required'}), 200

    # Validate parent_id to prevent negative values
    parent_id = data.get('parent_id')
    if parent_id is not None and parent_id < 0:
        return jsonify({'success': False, 'message': 'Invalid parent_id: cannot be negative'}), 200

    try:
        # Parse creation date if provided, otherwise use current date
        creation_date_str = data.get('creation_date')
        creation_date_dt = None
        if creation_date_str:
            try:
                # Treat the incoming date as local time, store it as-is
                creation_date_dt = datetime.strptime(creation_date_str, '%Y-%m-%d')
            except ValueError:
                return jsonify({'success': False, 'message': 'Invalid date format. Use YYYY-MM-DD'}), 200

        # Parse deadline if provided
        deadline_str = data.get('deadline')
        deadline_dt = None
        if deadline_str:
            try:
                # Store deadline in local time as well
                deadline_dt = datetime.fromisoformat(deadline_str)
            except ValueError:
                return jsonify({'success': False, 'message': 'Invalid deadline format'}), 200

        new_task = add_task(
            session=db.session,
            name=data['name'],
            user_id=user_id,
            description=data.get('description'),
            parent_id=data.get('parent_id'),
            category_id=data.get('category_id'),
            priority=data.get('priority'),
            creation_date=creation_date_dt,
            deadline=deadline_dt
        )

        # Get the task with its subtasks for the response
        task_data = get_task_with_subtasks(db.session, new_task.id, user_id)
        
        # Emit WebSocket event for real-time updates
        emit_task_created(task_data, user_id)

        return jsonify({
            'success': True,
            'data': task_data
        }), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'message': str(e)}), 500


@tasks_bp.route('/<int:task_id>', methods=['PUT'])
@jwt_required()
def update_task(task_id):
    user_id = get_jwt_identity()
    task = Task.query.filter_by(id=task_id, user_id=user_id).first()

    if not task:
        return jsonify({'error': 'Task not found'}), 404

    data = request.get_json()

    # If parent_id changes, move the subtask to a new parent
    if 'parent_id' in data:
        new_parent_id = data['parent_id']
        if not move_subtask(db.session, task_id, new_parent_id, user_id):
            return jsonify({'error': 'Failed to move task'}), 400
        # Refresh the task object after moving
        task = Task.query.filter_by(id=task_id, user_id=user_id).first()

    # Update other fields
    if 'name' in data:
        task.name = data['name']
    if 'description' in data:
        task.description = data['description']
    if 'completed' in data:
        task.completed = data['completed']
    if 'priority' in data:
        task.priority = data['priority']
    if 'category_id' in data:
        task.category_id = data['category_id']
    if 'deadline' in data:
        deadline_str = data['deadline']
        if deadline_str:
            try:
                task.deadline = datetime.fromisoformat(deadline_str)
            except ValueError:
                return jsonify({'error': 'Invalid deadline format'}), 400
        else:
            # Allow setting deadline to None/null
            task.deadline = None

    db.session.commit()

    # Get the updated task with its subtasks
    task_data = get_task_with_subtasks(db.session, task.id, user_id)
    
    # Emit WebSocket event for real-time updates
    emit_task_updated(task_data, user_id)

    # Return the updated task with the full subtask tree
    return jsonify(task_data)


@tasks_bp.route('/<int:task_id>', methods=['DELETE'])
@jwt_required()
def delete_task_route(task_id):
    user_id = get_jwt_identity()
    
    # Get the task's creation date before deleting it
    task = Task.query.filter_by(id=task_id, user_id=user_id).first()
    date_str = None
    if task and task.creation_date:
        date_str = task.creation_date.strftime('%Y-%m-%d')

    if delete_task(db.session, task_id, user_id):
        # Emit WebSocket event for real-time updates
        emit_task_deleted(task_id, user_id, date_str)
        return jsonify({'message': 'Task and all subtasks deleted successfully'}), 200
    else:
        return jsonify({'error': 'Failed to delete task'}), 400


@tasks_bp.route('/<int:task_id>/toggle', methods=['PUT'])
@jwt_required()
def toggle_task_completion_route(task_id):
    user_id = get_jwt_identity()
    
    # Use the utility function from task_utils.py
    result = toggle_task_completion(db.session, task_id, user_id)
    
    if not result:
        return jsonify({'error': 'Task not found'}), 404
    
    # Get the task's creation date and completed status
    task = Task.query.filter_by(id=task_id, user_id=user_id).first()
    date_str = None
    if task and task.creation_date:
        date_str = task.creation_date.strftime('%Y-%m-%d')
    
    # Emit WebSocket event for real-time updates
    emit_task_completed(task_id, task.completed, user_id, date_str)
        
    return jsonify(result)


@tasks_bp.route('/<int:task_id>/move', methods=['PUT', 'OPTIONS'])
@jwt_required()
def move_task_route(task_id):
    if request.method == 'OPTIONS':
        # Handle preflight request
        return '', 200

    user_id = get_jwt_identity()
    data = request.get_json()
    new_parent_id = data.get('parent_id')
    
    # Validate that we're not trying to move a task to be its own child
    if new_parent_id == task_id:
        return jsonify({'error': 'Cannot move a task to be its own child'}), 400

    result = move_subtask(db.session, task_id, new_parent_id, user_id)
    if not result:
        return jsonify({'error': 'Failed to move task. The task may not exist or would create a circular reference.'}), 400

    # Get the updated task with its subtasks
    task_data = get_task_with_subtasks(db.session, task_id, user_id)
    
    # Emit WebSocket event for real-time updates
    emit_task_updated(task_data, user_id)

    # Return the updated task with its subtasks
    return jsonify(task_data)


@tasks_bp.route('/search', methods=['GET'])
@jwt_required()
def search_tasks():
    """
    Search and filter tasks with comprehensive parameters.
    Query params:
    - time_scope: 'daily' | 'weekly' | 'monthly' | 'yearly'
    - anchor_date: YYYY-MM-DD (default: today)
    - client_today: YYYY-MM-DD (client's local today for timezone support)
    - search_query: text search
    - category_ids: comma-separated IDs
    - priority_levels: comma-separated levels
    - deadline_before: ISO date
    - deadline_after: ISO date
    - completion_status: 'all' | 'completed' | 'incomplete'
    """
    user_id = get_jwt_identity()
    
    # Get time scope and anchor date
    time_scope = request.args.get('time_scope', 'daily')
    anchor_date_str = request.args.get('anchor_date')
    client_today_str = request.args.get('client_today')
    
    try:
        # Parse anchor date or use today
        if anchor_date_str:
            anchor_date = datetime.strptime(anchor_date_str, '%Y-%m-%d')
        else:
            anchor_date = datetime.now()
        
        # Use client's today if provided, otherwise fall back to server's today
        if client_today_str:
            client_today = datetime.strptime(client_today_str, '%Y-%m-%d').date()
        else:
            client_today = datetime.now().date()
        
        # Calculate date range based on time scope
        from datetime import timedelta
        
        if time_scope == 'daily':
            start_date = anchor_date.replace(hour=0, minute=0, second=0, microsecond=0)
            end_date = anchor_date.replace(hour=23, minute=59, second=59, microsecond=999999)
        elif time_scope == 'weekly':
            # Get start of week (Monday)
            days_from_monday = anchor_date.weekday()
            start_date = (anchor_date - timedelta(days=days_from_monday)).replace(hour=0, minute=0, second=0, microsecond=0)
            end_date = (start_date + timedelta(days=6)).replace(hour=23, minute=59, second=59, microsecond=999999)
        elif time_scope == 'monthly':
            # Get start and end of month
            start_date = anchor_date.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
            # Get last day of month
            if anchor_date.month == 12:
                end_date = anchor_date.replace(month=12, day=31, hour=23, minute=59, second=59, microsecond=999999)
            else:
                next_month = anchor_date.replace(month=anchor_date.month + 1, day=1)
                end_date = (next_month - timedelta(days=1)).replace(hour=23, minute=59, second=59, microsecond=999999)
        elif time_scope == 'yearly':
            # Get start and end of year
            start_date = anchor_date.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
            end_date = anchor_date.replace(month=12, day=31, hour=23, minute=59, second=59, microsecond=999999)
        else:
            return jsonify({'error': 'Invalid time_scope. Use: daily, weekly, monthly, or yearly'}), 400
        
        # Get filter parameters
        search_query = request.args.get('search_query')
        category_ids_str = request.args.get('category_ids')
        priority_levels_str = request.args.get('priority_levels')
        deadline_before_str = request.args.get('deadline_before')
        deadline_after_str = request.args.get('deadline_after')
        completion_status = request.args.get('completion_status', 'all')
        
        # Parse filter parameters
        category_ids = None
        if category_ids_str:
            category_ids = [int(id.strip()) for id in category_ids_str.split(',') if id.strip()]
        
        priority_levels = None
        if priority_levels_str:
            priority_levels = [level.strip() for level in priority_levels_str.split(',') if level.strip()]
        
        deadline_before = None
        if deadline_before_str:
            deadline_before = datetime.fromisoformat(deadline_before_str)
        
        deadline_after = None
        if deadline_after_str:
            deadline_after = datetime.fromisoformat(deadline_after_str)
        
        # Get filtered tasks
        tasks = get_tasks_with_filters(
            session=db.session,
            user_id=user_id,
            start_date=start_date,
            end_date=end_date,
            search_query=search_query,
            category_ids=category_ids,
            priority_levels=priority_levels,
            deadline_before=deadline_before,
            deadline_after=deadline_after,
            completion_status=completion_status,
            client_today=client_today
        )
        
        return jsonify(tasks)
    except ValueError as e:
        return jsonify({'error': f'Invalid date format: {str(e)}'}), 400
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/stats', methods=['GET'])
@jwt_required()
def get_tasks_stats():
    user_id = get_jwt_identity()
    start_date = request.args.get('start_date')
    end_date = request.args.get('end_date')
    category_ids_str = request.args.get('category_ids')
    priority_levels_str = request.args.get('priority_levels')
    client_today_str = request.args.get('client_today')

    if not start_date or not end_date:
        return jsonify({'error': 'Both start_date and end_date are required'}), 400

    try:
        start_dt = datetime.strptime(start_date, '%Y-%m-%d')
        end_dt = datetime.strptime(end_date, '%Y-%m-%d')
        
        # Use client's today if provided, otherwise fall back to server's today
        if client_today_str:
            client_today = datetime.strptime(client_today_str, '%Y-%m-%d').date()
        else:
            client_today = datetime.now().date()
        
        # Parse category IDs if provided
        category_ids = None
        if category_ids_str:
            category_ids = [int(cid.strip()) for cid in category_ids_str.split(',') if cid.strip()]
        
        # Parse priority levels if provided
        priority_levels = None
        if priority_levels_str:
            priority_levels = [p.strip() for p in priority_levels_str.split(',') if p.strip()]
        
        # Get stats with optional filters
        stats = get_tasks_stats_by_date_range(
            db.session,
            user_id,
            start_dt,
            end_dt,
            category_ids=category_ids,
            priority_levels=priority_levels,
            client_today=client_today
        )
        
        return jsonify(stats)
    except ValueError:
        return jsonify({'error': 'Invalid date format. Use YYYY-MM-DD without timezone'}), 400
    except Exception as e:
        return jsonify({'error': str(e)}), 500


# Canvas View Endpoints

@tasks_bp.route('/<int:task_id>/position', methods=['POST'])
@jwt_required()
def update_task_position(task_id):
    """Update task position on canvas"""
    user_id = get_jwt_identity()
    task = Task.query.filter_by(id=task_id, user_id=user_id).first()

    if not task:
        return jsonify({'error': 'Task not found'}), 404

    data = request.get_json()
    
    # Validate that x and y are provided
    if 'x' not in data or 'y' not in data:
        return jsonify({'error': 'Both x and y coordinates are required'}), 400
    
    try:
        # Update position
        task.position_x = float(data['x'])
        task.position_y = float(data['y'])
        db.session.commit()
        
        return jsonify({
            'success': True,
            'data': {
                'id': task.id,
                'position_x': task.position_x,
                'position_y': task.position_y
            }
        }), 200
    except (ValueError, TypeError) as e:
        db.session.rollback()
        return jsonify({'error': 'Invalid coordinate values'}), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/<int:task_id>/customize', methods=['POST'])
@jwt_required()
def customize_task(task_id):
    """Update task appearance (color, shape) on canvas"""
    user_id = get_jwt_identity()
    task = Task.query.filter_by(id=task_id, user_id=user_id).first()

    if not task:
        return jsonify({'error': 'Task not found'}), 404

    data = request.get_json()
    
    try:
        if 'color' in data:
            task.canvas_color = data['color']
        if 'shape' in data:
            task.canvas_shape = data['shape']
        
        db.session.commit()
        
        return jsonify({
            'success': True,
            'data': {
                'id': task.id,
                'canvas_color': task.canvas_color,
                'canvas_shape': task.canvas_shape
            }
        }), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/dependencies', methods=['GET'])
@jwt_required()
def get_dependencies():
    """Get all task dependencies for the current user"""
    user_id = get_jwt_identity()
    
    try:
        dependencies = get_user_dependencies(db.session, user_id)
        return jsonify({'success': True, 'data': dependencies}), 200
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/dependencies', methods=['POST'])
@jwt_required()
def create_dependency():
    """Create a new task dependency"""
    user_id = get_jwt_identity()
    data = request.get_json()
    
    source_id = data.get('source_task_id')
    target_id = data.get('target_task_id')
    
    if not source_id or not target_id:
        return jsonify({'error': 'Both source_task_id and target_task_id are required'}), 400
    
    try:
        dependency = add_dependency(db.session, source_id, target_id, user_id)
        return jsonify({
            'success': True,
            'data': dependency.to_dict()
        }), 201
    except ValueError as e:
        return jsonify({'error': str(e)}), 400
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/dependencies/<int:dependency_id>', methods=['DELETE'])
@jwt_required()
def delete_dependency(dependency_id):
    """Delete a task dependency"""
    user_id = get_jwt_identity()
    
    try:
        remove_dependency(db.session, dependency_id, user_id)
        return jsonify({
            'success': True,
            'message': 'Dependency deleted successfully'
        }), 200
    except ValueError as e:
        return jsonify({'error': str(e)}), 404
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/<int:task_id>/dependencies', methods=['GET'])
@jwt_required()
def get_task_dependencies_route(task_id):
    """Get all dependencies for a specific task"""
    user_id = get_jwt_identity()
    
    try:
        dependencies = get_task_dependencies(db.session, task_id, user_id)
        return jsonify({'success': True, 'data': dependencies}), 200
    except ValueError as e:
        return jsonify({'error': str(e)}), 404
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@tasks_bp.route('/dependencies/<int:dependency_id>/customize', methods=['PUT'])
@jwt_required()
def customize_dependency(dependency_id):
    """Update dependency edge appearance (color, style, width, animated)"""
    user_id = get_jwt_identity()
    
    from models.task_dependency import TaskDependency
    
    dependency = db.session.query(TaskDependency).filter_by(
        id=dependency_id,
        user_id=user_id
    ).first()
    
    if not dependency:
        return jsonify({'error': 'Dependency not found'}), 404
    
    data = request.get_json()
    
    try:
        # Update edge customization fields
        if 'edge_color' in data:
            dependency.edge_color = data['edge_color']
        if 'edge_style' in data:
            # Validate edge style
            valid_styles = ['smoothstep', 'straight', 'step', 'bezier']
            if data['edge_style'] not in valid_styles:
                return jsonify({'error': f'Invalid edge style. Must be one of: {", ".join(valid_styles)}'}), 400
            dependency.edge_style = data['edge_style']
        if 'edge_width' in data:
            dependency.edge_width = float(data['edge_width'])
        if 'edge_animated' in data:
            dependency.edge_animated = bool(data['edge_animated'])
        
        db.session.commit()
        
        return jsonify({
            'success': True,
            'data': dependency.to_dict()
        }), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

File: client/src/services/tasks.ts
import api from './api';

// ------------------
// Types for API
// ------------------

/**
 * Request shape used to create or update a Task,
 * expecting an ISO string for `creation_date`.
 */
export interface TaskCreateRequest {
  name: string;
  category_id?: number;
  description?: string;
  priority?: string;
  parent_id: number | null;
  creation_date?: string | null;  // ISO string (if provided)
  deadline?: string | null;  // Optional ISO string for deadline
}

/**
 * The backend returns tasks in this shape, using ISO strings for dates.
 */
export interface TaskResponse {
  id: number;
  name: string;
  description?: string;
  completed: boolean;
  priority?: { color: string; level: string; } | string; // Support both new object format and legacy string format
  category_id?: number;
  parent_id?: number | null;
  creation_date: string;          // An ISO date string from the backend
  deadline?: string;             // Optional ISO string for deadline
  subtasks?: TaskResponse[];      // 'subtasks' are nested tasks
  category?: CategoryResponse;    // Additional info about the category
  position_x?: number | null;     // Canvas X position
  position_y?: number | null;     // Canvas Y position
  canvas_color?: string | null;   // Custom canvas color
  canvas_shape?: string | null;   // Canvas shape: 'rectangle', 'rounded', 'circle'
}

export interface CategoryResponse {
  id: number;
  name: string;
  description?: string;
  icon?: string;
}

export interface TaskDependency {
  id: number;
  source_task_id: number;
  target_task_id: number;
  user_id: number;
  edge_color?: string | null;
  edge_style?: string | null;
  edge_width?: number | null;
  edge_animated?: boolean | null;
}

/**
 * Front-end shape for a Task. Again, store dates as ISO strings.
 */
export interface Task {
  id: number;
  name: string;
  description?: string;
  completed: boolean;
  priority?: { color: string; level: string; } | string; // Support both new object format and legacy string format
  category_id?: number;
  parent_id: number | null;
  creation_date: string; // Keep as ISO string
  deadline?: string;    // Optional ISO string for deadline
  collapsed?: boolean;
  children: Task[]; 
  category?: string;
  categoryIcon?: string;
  position_x?: number | null;     // Canvas X position
  position_y?: number | null;     // Canvas Y position
  canvas_color?: string | null;   // Custom canvas color
  canvas_shape?: string | null;   // Canvas shape: 'rectangle', 'rounded', 'circle'
}

// ------------------
// Conversion helpers
// ------------------

/**
 * Recursively converts the backend's TaskResponse (which may have nested subtasks)
 * into the front-end Task structure with 'children'.
 */
export function mapTaskResponseToTask(taskRes: TaskResponse): Task {
  return {
    id: taskRes.id,
    name: taskRes.name,
    description: taskRes.description,
    completed: taskRes.completed,
    priority: taskRes.priority, 
    parent_id: taskRes.parent_id ? Number(taskRes.parent_id) : null,
    category_id: taskRes.category_id,
    creation_date: taskRes.creation_date,         // Keep the ISO string 
    deadline: taskRes.deadline,                   // Add deadline field
    collapsed: false,                             // Default UI preference
    children: (taskRes.subtasks || []).map(mapTaskResponseToTask),
    category: taskRes.category?.name,
    categoryIcon: taskRes.category?.icon,
    position_x: taskRes.position_x,
    position_y: taskRes.position_y,
    canvas_color: taskRes.canvas_color,
    canvas_shape: taskRes.canvas_shape
  };
}

// ------------------
// Task service
// ------------------

export const tasksService = {
  /**
   * Fetches all root tasks. Each task can contain nested 'subtasks'.
   */
  getTasks: async (): Promise<Task[]> => {
    try {
      const response = await api.get<TaskResponse[]>('/tasks/');
      const taskResponses = response.data;
      return taskResponses.map(mapTaskResponseToTask);
    } catch (error) {
      throw error;
    }
  },

  /**
   * Fetches tasks filtered by an exact ISO date/time string.
   * Example usage:
   *   tasksService.getTasksByDate(new Date().toISOString());
   */
  getTasksByDate: async (isoString: string): Promise<Task[]> => {
    try {
      // Ensure we're only sending the date part (YYYY-MM-DD) as the server expects
      const dateOnly = isoString.split('T')[0];
      const response = await api.get<TaskResponse[]>(`/tasks/?date=${encodeURIComponent(dateOnly)}`);
      const taskResponses = response.data;
      return taskResponses.map(mapTaskResponseToTask);
    } catch (error) {
      throw error;
    }
  },

  /**
   * Fetch a task's entire hierarchy by ID.
   */
  getTaskHierarchy: async (taskId: number) => {
    try {
      const response = await api.get(`/tasks/${taskId}/hierarchy`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Moves a task to a new parent, then returns the updated task.
   */
  moveTask: async (taskId: number, newParentId: number | null): Promise<Task> => {
    try {
      const response = await api.put(`/tasks/${taskId}/move`, { parent_id: newParentId });
      return mapTaskResponseToTask(response.data);
    } catch (error) {
      throw error;
    }
  },

  /**
   * Creates a new task. Make sure `creation_date` is an ISO string if provided.
   * If not provided, you can generate a default one here:
   *
   *   task.creation_date = task.creation_date ?? new Date().toISOString();
   */
  createTask: async (task: TaskCreateRequest): Promise<TaskResponse> => {
    try {
      // Optional: Provide a default date if none is supplied
      if (!task.creation_date) {
        // Use local date string in YYYY-MM-DD format to preserve local timezone
        const now = new Date();
        task.creation_date = `${now.getFullYear()}-${String(now.getMonth() + 1).padStart(2, '0')}-${String(now.getDate()).padStart(2, '0')}`;
      }
      
      // If deadline is provided as a full ISO string, keep it as is
      // The backend will handle it in local timezone
      
      const response = await api.post('/tasks/', task);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Updates an existing task. Partial updates can include a new ISO creation_date.
   */
  updateTask: async (taskId: number, updates: Partial<TaskCreateRequest>): Promise<TaskResponse> => {
    try {
      const response = await api.put(`/tasks/${taskId}`, updates);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Deletes a task by ID.
   */
  deleteTask: async (taskId: number): Promise<{ message: string }> => {
    try {
      const response = await api.delete(`/tasks/${taskId}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Toggles a task's 'completed' status on the server.
   * Some backends ignore the 'completed' value you send; it depends on the API.
   */
  toggleTaskComplete: async (taskId: number, _completed: boolean): Promise<TaskResponse> => {
    try {
      const response = await api.put(`/tasks/${taskId}/toggle`, {});
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Gets task statistics for a date range.
   * Returns counts of total and completed tasks for each day in the range.
   */
  getTaskStatsByDateRange: async (startDate: string, endDate: string): Promise<Array<{
    date: string;
    total_tasks: number;
    completed_tasks: number;
    completion_percentage: number;
  }>> => {
    try {
      const response = await api.get(`/tasks/stats?start_date=${encodeURIComponent(startDate)}&end_date=${encodeURIComponent(endDate)}`);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Adds a new subtask under a given parent task.
   */
  addSubtask: async (parentId: number, subtask: { name: string, parent_id?: number }): Promise<TaskResponse> => {
    try {
      // Always use the parentId parameter directly, ignoring any parent_id in the subtask object
      const taskData = {
        name: subtask.name,
        parent_id: parentId  // Use the parentId parameter directly
      };
      const response = await api.post('/tasks/', taskData);
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Updates just the name of a given task.
   */
  updateTaskName: async (taskId: number, newName: string): Promise<TaskResponse> => {
    try {
      const response = await api.put(`/tasks/${taskId}`, { name: newName });
      return response.data;
    } catch (error) {
      throw error;
    }
  },

  /**
   * Search tasks with comprehensive filters
   */
  searchTasksWithFilters: async (params: {
    timeScope: 'daily' | 'weekly' | 'monthly' | 'yearly';
    anchorDate: string;
    searchQuery?: string;
    categoryIds?: number[];
    priorityLevels?: string[];
    deadlineBefore?: string;
    deadlineAfter?: string;
    completionStatus?: 'all' | 'completed' | 'incomplete';
  }): Promise<Task[]> => {
    try {
      // Build query parameters
      const queryParams = new URLSearchParams();
      queryParams.append('time_scope', params.timeScope);
      queryParams.append('anchor_date', params.anchorDate);
      
      if (params.searchQuery) {
        queryParams.append('search_query', params.searchQuery);
      }
      if (params.categoryIds && params.categoryIds.length > 0) {
        queryParams.append('category_ids', params.categoryIds.join(','));
      }
      if (params.priorityLevels && params.priorityLevels.length > 0) {
        queryParams.append('priority_levels', params.priorityLevels.join(','));
      }
      if (params.deadlineBefore) {
        queryParams.append('deadline_before', params.deadlineBefore);
      }
      if (params.deadlineAfter) {
        queryParams.append('deadline_after', params.deadlineAfter);
      }
      if (params.completionStatus) {
        queryParams.append('completion_status', params.completionStatus);
      }

      const response = await api.get<TaskResponse[]>(`/tasks/search?${queryParams.toString()}`);
      const taskResponses = response.data;
      return taskResponses.map(mapTaskResponseToTask);
    } catch (error) {
      throw error;
    }
  },

  /**
   * Updates the position of a task on the canvas
   */
  updateTaskPosition: async (taskId: number, x: number, y: number): Promise<void> => {
    try {
      await api.post(`/tasks/${taskId}/position`, { x, y });
    } catch (error) {
      console.error('Failed to update task position:', error);
      throw error;
    }
  },

  /**
   * Updates the appearance (color, shape) of a task on the canvas
   */
  updateTaskAppearance: async (
    taskId: number,
    color?: string | null,
    shape?: string | null
  ): Promise<void> => {
    try {
      await api.post(`/tasks/${taskId}/customize`, { color, shape });
    } catch (error) {
      console.error('Failed to update task appearance:', error);
      throw error;
    }
  },

  /**
   * Get all task dependencies for the current user
   */
  getDependencies: async (): Promise<TaskDependency[]> => {
    try {
      const response = await api.get('/tasks/dependencies');
      return response.data.data || [];
    } catch (error) {
      console.error('Failed to fetch dependencies:', error);
      throw error;
    }
  },

  /**
   * Create a new task dependency
   */
  createDependency: async (sourceTaskId: number, targetTaskId: number): Promise<TaskDependency> => {
    try {
      const response = await api.post('/tasks/dependencies', {
        source_task_id: sourceTaskId,
        target_task_id: targetTaskId,
      });
      return response.data.data;
    } catch (error) {
      console.error('Failed to create dependency:', error);
      throw error;
    }
  },

  /**
   * Delete a task dependency
   */
  deleteDependency: async (dependencyId: number): Promise<void> => {
    try {
      await api.delete(`/tasks/dependencies/${dependencyId}`);
    } catch (error) {
      console.error('Failed to delete dependency:', error);
      throw error;
    }
  },

  /**
   * Get dependencies for a specific task
   */
  getTaskDependencies: async (taskId: number): Promise<{ outgoing: TaskDependency[], incoming: TaskDependency[] }> => {
    try {
      const response = await api.get(`/tasks/${taskId}/dependencies`);
      return response.data.data;
    } catch (error) {
      console.error('Failed to fetch task dependencies:', error);
      throw error;
    }
  },

  /**
   * Update the appearance of a dependency edge
   */
  updateDependencyAppearance: async (
    dependencyId: number,
    appearance: {
      edge_color?: string | null;
      edge_style?: string | null;
      edge_width?: number | null;
      edge_animated?: boolean | null;
    }
  ): Promise<TaskDependency> => {
    try {
      const response = await api.put(`/tasks/dependencies/${dependencyId}/customize`, appearance);
      return response.data.data;
    } catch (error) {
      console.error('Failed to update dependency appearance:', error);
      throw error;
    }
  },
};
